docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
```

Comment:
Why do you need `includeFrecent`? Where would we use this "top sites" without frecent sites? If you need that, you can just interact with the pinned sites storage directly.

Let's try to keep this as simple as possible, and only add such flags if they're clearly very necessary.

Reply:
We do not know how many top sites to query for, so the UI can request the number of top sites it wants to display and the remaining can be filled by frecent.

For example, if the UI can display 10 top site tiles, we have 6 pinned sites, and with `includeFrecent` true, the remaining 4 would come from the history storage.

We took the idea from the discussion [in this PR](https://github.com/mozilla-mobile/android-components/pull/7258#issuecomment-661465105) to fill the remaining with the frecent data. If you add a new pinned site, one less frecent would show up.

Reply:
On the Fenix side, we want to provide a user setting to disable frecent sites. So, fetching frecent sites can be optional if the users doesn't want it.

Reply:
Oh, I agree completely that remaining site slots should be filled out w/ frecents.

> we want to provide a user setting to disable frecent sites

Interesting! that makes sense then.